rosbag2/src/rosbag2/sequential_reader.cpp

```diff
@@ -14,21 +14,28 @@
 
 #include "rosbag2/sequential_reader.hpp"
 
+#include <cassert>
 #include <memory>
 #include <stdexcept>
 #include <string>
 #include <utility>
 #include <vector>
 
+#include "rcpputils/filesystem_helper.hpp"
 #include "rosbag2/info.hpp"
+#include "rosbag2/logging.hpp"
+
 
 namespace rosbag2
 {
 
 SequentialReader::SequentialReader(
   std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
-  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory)
-: storage_factory_(std::move(storage_factory)), converter_factory_(std::move(converter_factory)),
+  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory,
+  std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
```

Comment:
From what I see here, the only purpose of the `metadata_io` is to load the actual metadata once. I think it makes sense to consider passing in the metadata directly instead of the `io`. That potentially means the io handle to the metadata.yaml file is held open for the lifetime of this sequential reader instance, but also means I have to mock the complete io instead of just the metadata struct for testing.

Reply:
I worry that doing so may lead us back to the same problem we had before when we loaded the metadata from the storage plugin and we don't have up to date information.

While the reader should not be responsible for ever writing metadata, there may be a time where we need to reload the metadata (ex. streaming from bags that are also being written) and it would be helpful to have metadata_io read the updated file.

That, and this new constructor would mirror the writer nicely.

Reply:
makes sense to me.